pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,144 @@
+	// if the CRD does not exist (or is deleted)
+	mcsSupported bool
+}
+
+// ServiceExportOptions provide options for creating a ServiceExportController.
+type ServiceExportOptions struct {
+	Client       kube.Client
+	ClusterID    string
+	DomainSuffix string
+	ClusterLocal model.ClusterLocalProvider
+}
+
+// NewServiceExportController creates a new ServiceExportController.
+func NewServiceExportController(opts ServiceExportOptions) *ServiceExportController {
+	c := &ServiceExportController{
+		ServiceExportOptions: opts,
+		client:               opts.Client.MCSApis(),
+		serviceClient:        opts.Client.Kube().CoreV1(),
+		queue:                queue.NewQueue(time.Second),
+		mcsSupported:         true,
+	}
+
+	c.serviceInformer = opts.Client.KubeInformer().Core().V1().Services().Informer()
+	c.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) { c.onServiceAdd(obj) },
+		// Do nothing on delete. When we create ServiceExport, we bind its
+		// lifecycle to the Service so that when the Service is deleted,
+		// k8s automatically deletes the ServiceExport.
+	})
+
+	return c
+}
+
+func (c *ServiceExportController) onServiceAdd(obj interface{}) {
+	c.queue.Push(func() error {
+		if !c.mcsSupported {
+			// Don't create ServiceExport if MCS is not supported on the cluster.
+			return nil
+		}
+
+		svc, err := convertToService(obj)
+		if err != nil {
+			return err
+		}
+
+		if c.isClusterLocal(svc) {
+			// Don't create ServiceExport if the service is configured to be
+			// local to the cluster (i.e. non-exported).
+			return nil
+		}
+
+		return c.createServiceExportIfNotPresent(svc)
+	})
+}
+
+func (c *ServiceExportController) Run(stopCh <-chan struct{}) {
+	cache.WaitForCacheSync(stopCh, c.serviceInformer.HasSynced)
+	log.Infof("ServiceExport controller started")
+	go c.queue.Run(stopCh)
+}
+
+func (c *ServiceExportController) createServiceExportIfNotPresent(svc *v1.Service) error {
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = svc.Namespace
+	serviceExport.Name = svc.Name
+
+	// Bind the lifecycle of the ServiceExport to the Service. We do this by making the Service
+	// the "owner" of the ServiceExport resource.
+	serviceExport.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion: v1.SchemeGroupVersion.String(),
+			Kind:       "Service",
+			Name:       svc.Name,
+			UID:        svc.UID,
+		},
+	}
+
+	serviceExports := c.client.MulticlusterV1alpha1().ServiceExports(svc.Namespace)
+	_, err := serviceExports.Create(context.TODO(), &serviceExport, metav1.CreateOptions{})
+	if err != nil {
+		switch {
+		case errors.IsAlreadyExists(err):
+			// The ServiceExport already exists. Nothing to do.
+			return nil
+		case errors.IsNotFound(err):
+			log.Errorf("ServiceExport CRD Not found in cluster %s. Shutting down MCS ServiceExport sync. "+
+				"Please add the CRD then restart the istiod deployment", c.ClusterID)
+			c.mcsSupported = false
```

Comment:
this is race-y, maybe we should use atomic.Bool

Reply:
All of this is called in the queue thread, should all be single-threaded. Or am I missing something?

Reply:
nope, I was the one missing it :slightly_smiling_face: LGTM